src/hooks/useSwapSlippageTolerance.ts

```diff
@@ -1,25 +1,61 @@
-import { Currency, Percent, TradeType } from '@uniswap/sdk-core'
+import { Currency, CurrencyAmount, Percent, TradeType } from '@uniswap/sdk-core'
 import { Trade as V2Trade } from '@uniswap/v2-sdk'
 import { Trade as V3Trade } from '@uniswap/v3-sdk'
 import { L2_CHAIN_IDS } from 'constants/chains'
+import JSBI from 'jsbi'
 import { useMemo } from 'react'
 
 import { useUserSlippageToleranceWithDefault } from '../state/user/hooks'
+import { useCurrency } from './Tokens'
+import useGasPrice from './useGasPrice'
+import useUSDCPrice, { useUSDCValue } from './useUSDCPrice'
 import { useActiveWeb3React } from './web3'
 
 const V2_SWAP_DEFAULT_SLIPPAGE = new Percent(50, 10_000) // .50%
 const V3_SWAP_DEFAULT_SLIPPAGE = new Percent(50, 10_000) // .50%
 const ONE_TENTHS_PERCENT = new Percent(10, 10_000) // .10%
 
+/**
+ * Return a guess of the gas cost used in computing slippage tolerance for a given trade
+ * @param trade the trade for which to _guess_ the amount of gas it would cost to execute
+ */
+function guestimateGas(
+  trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType> | undefined
+): number | undefined {
+  if (!trade) return undefined
+  if (trade instanceof V2Trade) return 90_000 + trade.route.pairs.length * 20_000
+  // does not consider the number of ticks crossed
+  return 100_000 + trade.swaps.reduce((memo, swap) => swap.route.pools.length + memo, 0) * 30_000
+}
+
 export default function useSwapSlippageTolerance(
   trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType> | undefined
 ): Percent {
   const { chainId } = useActiveWeb3React()
   const onL2 = chainId && L2_CHAIN_IDS.includes(chainId)
+  const outputDollarValue = useUSDCValue(trade?.outputAmount)
+  const ethGasPrice = useGasPrice()
+
+  const gasEstimate = guestimateGas(trade)
+  const ether = useCurrency('ETH')
+  const etherPrice = useUSDCPrice(ether ?? undefined)
+
   const defaultSlippageTolerance = useMemo(() => {
     if (!trade || onL2) return ONE_TENTHS_PERCENT
+
+    const ethGasCost =
+      ethGasPrice && typeof gasEstimate === 'number' ? JSBI.multiply(ethGasPrice, JSBI.BigInt(gasEstimate)) : undefined
+    const dollarGasCost =
+      ether && ethGasCost && etherPrice ? etherPrice.quote(CurrencyAmount.fromRawAmount(ether, ethGasCost)) : undefined
+
+    if (outputDollarValue && dollarGasCost) {
+      const fraction = dollarGasCost.asFraction.divide(outputDollarValue.asFraction)
```

Comment:
Comment this logic. What are you trying to set? Slippage can be up to or equal to gas cost?